File: src/main/java/parser/Parser.java
package parser;

import ast.Ast;
import ast.Node;
import scanner.Scanner;
import scanner.Token;

import java.util.Arrays;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * A parser converts a scanner to a tree structure.
 *
 * @param <T> The {@link Token} object.
 * @param <E> The {@link Token}s type.
 * @param <O> The {@link SymbolTable}s members type.
 * @param <N> The {@link Node}s type.
 */
public abstract class Parser<T extends Token<E>, E extends Enum, O extends ParserObject, N extends Node<?, ?, ?>> {

    /**
     * The logger.
     */
    private static final Logger LOGGER = Logger.getLogger(Parser.class.getName());

    /**
     * The AST.
     */
    protected final Ast<N> ast;

    /**
     * The scanner.
     */
    private final Scanner<T, E> scanner;

    /**
     * The current token.
     */
    protected T token;

    /**
     * The symbol table.
     */
    protected SymbolTable<O> symbolTable;

    /**
     * Create a new parser.
     *
     * @param scanner The scanner.
     */
    public Parser(final Scanner<T, E> scanner) {
        this.scanner = scanner;
        this.ast = new Ast<>();
    }

    /**
     * Set current to the next {@link Token}.
     */
    protected void next() {
        this.token = scanner.next();
        if (token != null) {
            LOGGER.fine(this.token.toString());
        }
    }

    /**
     * Run the given code with its own {@link SymbolTable}.
     *
     * @param runnable The code to run.
     */
    protected void scope(Runnable runnable) {
        final SymbolTable<O> oldSymbolTable = symbolTable;
        symbolTable = new SymbolTable<>(oldSymbolTable);
        runnable.run();
        symbolTable = oldSymbolTable;
    }

    /**
     * Switch to the error state.
     *
     * @param expected A list of expected tokens.
     */
    protected void error(final List<E> expected) {
        String message = "Unexpected token " + token;

        if (expected.size() > 0) {
            message += ". Expected token of one of the following types: " + expected.toString() + ".";
        }

        LOGGER.log(Level.SEVERE, message);
        throw new RuntimeException();
    }

    /**
     * Start the parsing process (by calling the start node method).
     *
     * @return The generated {@link Ast}.
     */
    public abstract Ast<N> parse();

    /**
     * Allow verifications on the current token.
     *
     * @return A {@link TokenVerification} object.
     */
    protected TokenVerification token(final List<E> expected) {
        return new TokenVerification(expected);
    }

    /**
     * Allow verifications on the current token.
     *
     * @return A {@link TokenVerification} object.
     */
    @SafeVarargs
    protected final TokenVerification token(final E... expected) {
        return token(Arrays.asList(expected));
    }

    /**
     * As part of the specification the counter verifies the number of {@link Token}s to match.
     */
    protected class TokenVerification {

        /**
         * The expected tokens.
         */
        private final List<E> expected;

        /**
         * Create a new counter expecting the given {@link Token}s.
         *
         * @param expected The expected {@link Token}s.
         */
        private TokenVerification(final List<E> expected) {
            this.expected = expected;
        }

        /**
         * Make sure the token is available exactly once. Throw an error otherwise.
         */
        public T once() {
            if (expected.contains(token.getType())) {
                final T result = token;
                next();
                return result;
            } else {
                error(expected);
                return null;
            }
        }

        /**
         * If the {@link Token} matches the current {@link Token} execute the function.
         *
         * @param function The function.
         */
        public void optional(final Runnable function) {
            if (expected.contains(token.getType())) {
                function.run();
            }
        }

        /**
         * Repeat the function while the {@link Token} matches the current {@link Token}.
         *
         * @param function The function.
         */
        public void repeat(final Runnable function) {
            while (expected.contains(token.getType())) {
                function.run();
            }
        }
    }
}

File: src/main/scala/compiler/parser/Parser.scala
package compiler.parser

import java.util.NoSuchElementException

import com.typesafe.scalalogging.Logger
import compiler.ast.{Ast, Node}
import compiler.scanner.{Scanner, Token}
import org.slf4j.LoggerFactory

/**
  * A parser converts a scanner to a tree structure.
  *
  * @param scanner The [[compiler.scanner.Scanner]].
  * @tparam T The [[compiler.scanner.Token]] object.
  * @tparam E The [[compiler.scanner.Token]]s type.
  * @tparam O The [[SymbolTable]]s members type.
  * @tparam N The [[compiler.ast.Node]]s type.
  * @author Julian Schlichtholz
  * @version 1.0.0
  */
abstract class Parser[T <: Token[E], E <: Enumeration#Value, O <: ParserObject, N <: Node[_, _]](val scanner: Scanner[T, E]) {

  /**
    * The logger.
    */
  private val logger: Logger = Logger(LoggerFactory.getLogger(getClass))

  /**
    * The symbol table.
    */
  var symbolTable: SymbolTable[O, E] = new SymbolTable()

  /**
    * The current token.
    */
  protected var current: T

  /**
    * Start the parsing process (by calling the start node method).
    *
    * @return The generated [[compiler.ast.Ast]].
    */
  def parse: Ast[N]

  /**
    * Set current to the next [[compiler.scanner.Token]].
    */
  protected def next() {
    try {
      this.current = scanner.next()
    } catch {
      case e: NoSuchElementException => logger.error(s"Premature end of file: ${scanner.input.source}.", e)
    }
    if (this.current != null) {
      logger.debug(this.current.toString)
    }
  }

  /**
    * Run the given code with its own [[SymbolTable]].
    *
    * @param f The code to run.
    */
  protected def scope(f: () => Unit) {
    val oldSymbolTable: SymbolTable[O, E] = symbolTable
    symbolTable = new SymbolTable(Option(oldSymbolTable))
    f()
    symbolTable = oldSymbolTable
  }

  /**
    * Switch to the error state.
    *
    * @param expected An expected [[compiler.scanner.Token]].
    */
  protected def error(expected: E): Nothing = {
    error(Seq(expected))
  }

  /**
    * Switch to the error state.
    *
    * @param expected A list of expected tokens.
    */
  protected def error(expected: Seq[E]): Nothing = {
    var message: String = "Unexpected token " + current
    if (expected.nonEmpty) {
      message += ". Expected token of one of the following types: " + expected.mkString(", ") + "."
    }
    logger.error(message)
    throw new RuntimeException(message)
  }

  /**
    * Allow verifications on the current token.
    *
    * @return A [[TokenVerification]] object.
    */
  protected def token(expected: E): TokenVerification = {
    token(Seq(expected))
  }

  /**
    * Allow verifications on the current token.
    *
    * @return A [[TokenVerification]] object.
    */
  protected def token(expected: Seq[E]): TokenVerification = {
    new TokenVerification(expected)
  }

  /**
    * As part of the specification the counter verifies the number of [[compiler.scanner.Token]]s to match.
    *
    * @param expected The expected [[compiler.scanner.Token]]s.
    */
  class TokenVerification(val expected: Seq[E]) {

    /**
      * Make sure the token is available once. Throw an error otherwise.
      */
    def once: T = {
      if (expected.contains(current.typ)) {
        val result: T = current
        next()
        result
      } else {
        error(expected)
      }
    }

    /**
      * If the [[compiler.scanner.Token]] matches the current [[compiler.scanner.Token]] execute the function.
      *
      * @param f The function.
      */
    def optional[U](f: (T) => U) {
      if (expected.contains(current.typ)) {
        f(current)
      }
    }

    /**
      * Repeat the function while the [[compiler.scanner.Token]] matches the current [[compiler.scanner.Token]].
      *
      * @param f The function.
      */
    def repeat[U](f: () => U) {
      while (expected.contains(current.typ)) {
        f()
      }
    }
  }

}
